### PR TITLE
Optional spectral width threshold for constraining equatorward extent of HMB

### PR DESCRIPTION
**Pull request by pasha-ponomarenko**
This pull request introduces an optional spectral width threshold which allows to mitigate an issue with HMB being placed at unphysically low magnetic latitudes due to the presence of sub-auroral echoes with velocities unrelated to the high-latitude plasma circulation directly driven by the solar wind.  This issue is described in detail in this talk from SD'22 [Ponomarenko_On_utilising_spectral_width.pdf](https://github.com/SuperDARN/rst/files/8901793/Ponomarenko_On_utilising_spectral_width.pdf)

The threshold can be set by using a new switch in `map_addhmb.c` routine:
`map_addhmb -wdt 100 [input file] > [output file]`

By default the threshold value is set to 0 m/s so that no width-based filtering is performed.

Testing this option consists of processing the same data with and without the threshold and comparing the resulting HMB locations. For example, here is an example when during a very quiet interval (Kp=1-) the 2:42-2:44UT map (left) shows the expected extent of the HMB > 72 MLAT while in the following map 2:44-2:46UT (right) it has been pulled down to a very low MLAT ~50 MLAT by the sub-auroral echoes from BKS:

image

The plot below contains the same data but after application of the spectral width threshold of 100 m/s. As one can see, the HMB in the right map is now located at the expected MLAT > 70 deg:

image

### CODE REVIEW

**Comment by egthomas** [comment]
I haven't had a chance to look at this in any kind of detail yet, so just leaving two quick comments for now:

1. I changed the base branch from `main` to `develop` to avoid any accidental commits to the default branch
2. By default, spectral width information is not stored in `grid`-format files - users have to include the `-xtd` option when executing `make_grid`.  Do you know what happens if someone tries applying the optional `-wdt` threshold to a file which does not contain spectral width values?

**Comment by pasha-ponomarenko** [comment]
Good catch, @egthomas ! ...
If one applies the threshold to the data gridded without `-xdt` switch, the maps look interesting as some extra filtering is still performed 😎 , but I need more time to find out what actually happens and why.

**Comment by egthomas** [comment]
@pasha-ponomarenko the commit I just made should address that issue - it was similar to the problem of certain model vector fields being uninitialized which we fixed in #442.

**Comment by pasha-ponomarenko** [comment]
@egthomas, I have downloaded the corrected branch and tested it without `-xtd`. I realised that with the width threshold logics implemented in map_addhmb.c it makes sense to set the median width value to a very large number otherwise with any threshold value exceeding zero all gridded data are ignored while determining the HMB location, and HMB is placed at a (default?) latitude of 62 deg. I tested the code with wdt.median= 10000 and the width threshold of 100 m/s, and it worked as expected, i.e., produced the same results as for a zero threshold.
I am not sure what would be the best candidate for the very large number though... Is it 999999?

**Comment by pasha-ponomarenko** [comment]
@egthomas , @ecbland , another way to fix this issue would be to make writing median width into gridded files a default option. I don't think it takes too much space. 
What do you think?

File: codebase/superdarn/src.bin/tk/tool/map_addhmb.1.17/map_addhmb.c [hunk 1]
```diff
@@ -295,7 +299,8 @@ int main(int argc,char *argv[])
 
         for (j=0; j<nlat; j++) latcnt[j] = 0;
         for (i=0; i<grd[buf]->vcnum; i++) {
-          if (fabs(grd[buf]->data[i].vel.median) < vel_min) continue;
+        
+          if (fabs(grd[buf]->data[i].vel.median) < vel_min || grd[buf]->data[i].wdt.median < wdt_min) continue;
```
Comment (egthomas) [inline codebase/superdarn/src.bin/tk/tool/map_addhmb.1.17/map_addhmb.c:303]:
```suggestion
          if ( (fabs(grd[buf]->data[i].vel.median) < vel_min) ||
               ((grd[buf]->xtd) && (grd[buf]->data[i].wdt.median < wdt_min)) ) continue;
```

@pasha-ponomarenko would something like this solve the present issue?  It seems like requiring the calculation of spectral width values for all grid files is a more significant change than what is being proposed in this pull request.

(note there may also need to be a `grd->xtd=0;` inserted between lines 352-353 of `readmap.c` to ensure the `xtd` flag is being properly set)

Comment (pasha-ponomarenko) [reply codebase/superdarn/src.bin/tk/tool/map_addhmb.1.17/map_addhmb.c:303]: Thanks @egthomas , it looks fine for me!

**Comment by egthomas** [comment]
@pasha-ponomarenko ok, I think it should be working now.  I also fixed an issue where the `xtd` flag would never get reset to zero if for some reason records alternated between having and not having the extended power and spectral width information, and also changed some of the `CnvMapRead` logic to match `GridRead` when initializing the extended fields with zeros.

**Comment by pasha-ponomarenko** [comment]
Good work @egthomas !

**Comment by ecbland** [comment]
@pasha-ponomarenko @egthomas Would you be happy if I do the testing for this PR (noting that the first commit was mine)?

Or if someone else would like to test this PR, please leave a comment here!

**Comment by pasha-ponomarenko** [comment]
@ecbland , I have no issues withthis!

**Comment by egthomas** [comment]
@ecbland that's no problem, but maybe @mtwalach (or Adrian?) might want to take a look too since they were also doing HMB stuff recently?

**Comment by mtwalach** [comment]
Sorry I have been so absent, I will test this now.

**Comment by mtwalach** [comment]
I got as far as pulling this branch, 'make.build' and 'make.code'. 'make.code' failed on compiling the following binary: 

`================================================================================
Compiling Binary:/Users/walach/rst/codebase/superdarn/src.bin/tk/tcpip/fitacfclientgui.1.0
================================================================================
make clean
rm -f *.o
rm -f version.h
rm -f errstr.h
rm -f hlpstr.h
rm -f fitacfclientgui
make
make.version /Users/walach/rst/codebase/superdarn/src.bin/tk/tcpip/fitacfclientgui.1.0
make.help
./doc/fitacfclientgui.doc.xml
cc -Wall -pedantic -O3 -D_GNU_SOURCE -D_DARWIN -I/opt/local/include -D_SVGLIB_ -I/Users/walach/rst/include/base -I/Users/walach/rst/include/general -I/Users/walach/rst/include/superdarn -c fitacfclientgui.c    
mkdir -p /Users/walach/rst/bin
cc -Wl,-rpath,@loader_path/../lib/ -Wl,-rpath,@loader_path/../../lib/ -L/opt/local/lib -L/Users/walach/rst/lib -o /Users/walach/rst/bin/fitacfclientgui fitacfclientgui.o -lfitcnx.1 -lfit.1 -lcfit.1 -lrscan.1 -lradar.1 -lcnx.1 -ldmap.1 -lopt.1 -lrtime.1 -lrcnv.1 -lncurses -ltinfo  \
                            -lz 
ld: library not found for -ltinfo
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [fitacfclientgui] Error 1
Compilation Aborted.
`
This is a new one for me - and probably nothing to do with this branch, but rather how out of date the RST on this machine is. 
Any ideas?

**Comment by egthomas** [comment]
@mtwalach thanks for identifying this issue - it looks like you're using a Mac, correct?  Some quick googling suggests you may need to install or update your ncurses and/or libtinfo libraries.

**Comment by ecbland** [comment]
@mtwalach Thanks for having a look at this PR 💯 

What OS are you using? I've just checked and I can compile this branch ok on my machine (Linux Mint 20).

It looks like the new `fitacfclientgui` is the only binary that uses `-ltinfo`, so you probably can't compile `develop` either. @egthomas, do you know what package might need to be installed to fix this?

As a workaround to test this PR, you could remove the `rst/codebase/superdarn/src.bin/tk/tcpip/fitacfclientgui.1.0` directory and then try compiling again.

**Comment by ecbland** [comment]
Oh, sorry @egthomas, I missed your reply while writing mine :)

**Comment by mtwalach** [comment]
Hi, yes I am on a mac. Currently running Monterey V12.2. Thanks, I'll try to update these.

**Comment by pasha-ponomarenko** [comment]
@mtwalach, appreciate your help!

**Comment by mtwalach** [comment]
Hi again, 
After some reinstalling and a lot of failures due to outdated software, I got it to compile in the end by deleting the directory as Emma suggested. Then I got some grid files ready to process and off it went. It went really slowly and form the command line output, it was generating sensible HMBs BUT the file is now too huge to open - when I try to trim it, something corrupts because it generates an empty file but I get no error message with -vb, so I don't know what happens. 
I realised though that none of my grid files have the spectral width stored in them, so whilst it looks like it's generating something sensible, I don't actually know if it is..And like I said, I can't open the files. Since it will take me a bit of time to reprocess some fitacfs (I am travelling over the next few days - sorry folks), could someone send me some grid files with the spectral widths? @pasha-ponomarenko maybe you can send me the ones you used for the diagrams above and I can see if I can recreate the same outputs? Would love to do a more involved test (and I will at some point because I am curious about the differences in the maps) but it won't be instant.

**Comment by pasha-ponomarenko** [comment]
@mtwalach, please check your mailbox -- I sent you both ZIP and TAR.GZ with the requested GRD files. Please let me know if you can open those. Good luck! :-)

**Comment by egthomas** [comment]
@pasha-ponomarenko thanks for providing some example files!

@mtwalach the code may have been going really slowly on this branch because we hadn't yet merged in the changes from #511, which dramatically improved the speed of `map_addhmb`.  I've just pushed those updates to this branch, so hopefully it will process more quickly for you.

**Comment by mtwalach** [comment]
Hi, I got it working. Interestingly, I don't get the same results as @pasha-ponomarenko and I don't know why. 

Left is with spectral width (see code below) and right is with -vel 100 -cnt 3. 
image

image


Here is what I did: 
Combine all grids which Pasha sent me into one grid file: 
`combine_grid ./grd4maria/*grd > pasha.combined.grd` 
Made a copy: 
`cp pasha.combined.grd ./grid_file1.temp`
Fitting with the spectral width: 
`map_grd grid_file1.temp | map_addhmb -vb -wdt 100 | map_addmodel -o 8 -d l -ts18 | map_fit > test.fit.cnvmap`
Made a map file: 
`test.fit.cnvmap > tst1.fit.map`
Then the same for a HMB of `-vel 100 -cnt 3`:
`map_grd grid_file1.temp | map_addhmb -vb -vel 100 -cnt 3 | map_addmodel -o 8 -d l -ts18 | map_fit > test.fit.cnvmap
test.fit.cnvmap > tst2.fit.map`
Then I read them in and plotted them.

Interestingly, the hmb latitudes that come out when I make the maps with the -vb keyword are the same for both cases. Where am I going wrong?

**Comment by ecbland** [comment]
@mtwalach Thanks for testing and for identifying some unexpected behaviour :open_mouth: 
Would you mind sending the sample file to me so that I can try to reproduce the problem?

**Comment by mtwalach** [comment]
Hi @ecbland (and everyone else who wants to have a go)

Attached is the folder of grid files Pasha sent me (I hope you don't mind me sharing these here @pasha-ponomarenko) and the combined grid file I made from them.. I also attached my final map files. Please let me know if you have any problems. 
[grd4maria.zip](https://github.com/SuperDARN/rst/files/9850842/grd4maria.zip)
[pasha.combined.grd.zip](https://github.com/SuperDARN/rst/files/9850843/pasha.combined.grd.zip)
[tst2.fit.map.zip](https://github.com/SuperDARN/rst/files/9850858/tst2.fit.map.zip)
[tst1.fit.map.zip](https://github.com/SuperDARN/rst/files/9850859/tst1.fit.map.zip)

**Comment by pasha-ponomarenko** [comment]
@mtwalach, it might be that I have sent you GRD data without spectral width as I we were looking at a fix for non-width GRD files that was implemented by Evan in August (see above).  Apologies for this! 
:-( 
I will try to regenerate the same files with median spectral width ASAP.

**Comment by mtwalach** [comment]
Ah, I just assumed they were there but didn't take a closer look. That would explain it all.. :)

**Comment by pasha-ponomarenko** [comment]
@egthomas, could we have some sort of warning in ADD_HMB if the width threshold is applied to GRD data with no spectral width?

**Comment by egthomas** [comment]
@pasha-ponomarenko that could be done, although it is complicated by the fact that one can create a grid file by combining any number of other grid files, each of which may or may not contain spectral width information.  The easiest solution may be to just check whether the first record of the input grid file or stream contains spectral width and print the warning once.

**Comment by mtwalach** [comment]
@pasha-ponomarenko Please can you send this again, I don't seem to have received this..

**Comment by pasha-ponomarenko** [comment]
@mtwalach, re-sent them from my Windows machine as the Linux mailbox seems to play games... Please confirm.

**Comment by pasha-ponomarenko** [comment]
Just uploaded zipped files here:
[grd_ind.zip](https://github.com/SuperDARN/rst/files/9853754/grd_ind.zip)

**Comment by mtwalach** [comment]
Good news everyone, going through the same steps again with Pasha's other set of files, I made the following (set-up is the same as previously described, except now I get what we expect):

image

image

Happy for this to be added now, though as Pasha suggested, a warning when no spectral width is available would be useful. Surely it should only work when all measurements have spectral widths associated with them? Otherwise it would  omit some data and that would not be an improvement in my opinion. 

I will do larger-scale testing at some later point but it's otherwise probably good to go :) Sorry it took so long!

**Review by mtwalach — APPROVED** [review]

**Comment by pasha-ponomarenko** [comment]
@ecbland, should we ask Maria to press the green button?

**Comment by mtwalach** [comment]
I would love to press the green button. Can I press the green button? 😮 On github the green button acts like a red button..

**Comment by ecbland** [comment]
> I would love to press the green button. Can I press the green button? open_mouth On github the green button acts like a red button..

Go for it @mtwalach !!

**Comment by egthomas** [comment]
Just as a follow-up, we're currently testing this new spectral width option (`map_addhmb -nodef -wdt 100`) with our real-time convection display here:

http://superdarn.thayer.dartmouth.edu/

My initial impression is that a `-cnt` threshold may also need to be applied but that isn't based on any kind of quantitative analysis.

**Comment by pasha-ponomarenko** [comment]
@egthomas, is the default value of cnt_req = 3 not working?  Also, do you actually need to use -nodef switch?

**Comment by egthomas** [comment]
@pasha-ponomarenko sorry, I meant maybe the `-cnt` threshold should be increased from the default value of 3.  Previously I think we were using just the default velocity criteria with a (slightly increased) threshold of 5 vectors for our real-time display.

I think the `-nodef` option just tells Map Potential to use the HMB of the background model pattern if no real data vectors are available; otherwise, it will always use a fixed HMB of 62(?) degrees regardless of the IMF conditions.

**Comment by pasha-ponomarenko** [comment]
@egthomas, I am not sure about that as from your RT display this afternoon the HMB location sometimes went to very low latitudes < 50 MLAT that may not be coming from the statistical model.  Did you try to not use `-nodef`?

**Pull request merged**